unixfs/io/pbdagreader.go

```diff
@@ -131,8 +141,18 @@ func (dr *PBDagReader) precalcNextBuf(ctx context.Context) error {
 	default:
 		var err error
 		dr.buf, err = NewDagReader(ctx, nxt, dr.serv)
-		return err
+		if err != nil {
+			return err
+		}
 	}
+	if len(dr.pbdata.Blocksizes) > 0 {
+		// if blocksizes are set, ensure that we read exactly that amount of data
+		dr.buf, err = newSizeAdjReadSeekCloser(dr.buf, dr.pbdata.Blocksizes[linkPos])
```

Comment:
You know, I think we can do something simpler here. Given that we *know* the Filesize of the buffer, we can require that `buf` export a `Size()` method. Then, we can just have a check here:

* `buf.Size() > blocksize` -> make a limit seeker
* `==` -> return as-is
* `<` -> zero extend

That way, we can do all of these checks up-front. Thoughts? Too late?